File: src/main/problems/java/Nr5.java
package main.problems.java;

public class Nr5 {

    public static void main(String[] args) {
        Nr5 nr5 = new Nr5();
        System.out.println(nr5.longestPalindrome("cbbd"));
    }

    public String longestPalindrome(String s) {
        if (s == null || s.length() == 1) {
            return s;
        }
        int start = 0, end = 0;

        for (int i = 0; i < s.length(); i++) {
            int nonLetterCenter = expandAroundCenter(s, i, i + 1);
            int letterCenter = expandAroundCenter(s, i, i);
            int maxLength = Math.max(nonLetterCenter, letterCenter);
            if (maxLength > end - start) {
                start = i - (maxLength - 1) / 2;
                end = i + maxLength / 2;
            }
        }

        return s.substring(start, end + 1);
    }

    private int expandAroundCenter(String s, int left, int right) {
        while (
                left >= 0
                        && right < s.length()
                        && s.charAt(left) == s.charAt(right)
        ) {
            left--;
            right++;
        }
        return right - left - 1;
    }
}

File: src/main/problems/ruby/Nr5.rb
# @param {String} s
# @return {String}
def longest_palindrome(s)
    left, right = 0, 0
    (0...s.size).each do |idx|
        expanded_left, expanded_right = expand(s, idx, idx)
        left, right = expanded_left, expanded_right if (expanded_right - expanded_left) > (right - left)
        if idx < s.size - 1
            expanded_left, expanded_right = expand(s, idx + 1, idx)
            left, right = expanded_left, expanded_right if (expanded_right - expanded_left) > (right - left)
        end
    end
    s[left..right]
end

def expand(s, left, right)
    while left > 0 && right < s.size - 1 && s[left - 1] == s[right + 1]
      left -= 1
      right += 1
    end
    return left, right
end